### ISSUE TITLE
Catapult trybot are stuck

### ISSUE THREAD

**Issue by nedn** [catapult-project/catapult#2254]
https://build.chromium.org/p/tryserver.client.catapult/builders/Catapult%20Mac%20Tryserver/builds/2855

**Comment by nedn** [comment]
I think this is caused by my CL in https://codereview.chromium.org/1893623002/ but someone who is knowledgeable with catapult bot should restart the tasks.

**Comment by nedn** [comment]
My tasks are cancelled now, so downgrade this issue. However, we should still implement timeout in catapult recipe.

**Comment by anniesullie** [comment]
@riannucci - Any ideas how we can make sure our generator_script recipe has some kind of timeouts set on commands? This build ran 4 hours before we killed it, I think on a per-test basis we'd want closer to a 20 minute timeout.

**Comment by riannucci** [comment]
In the master config you can set a maximum timeout (and a max timeout w/o output). It's probably set way too high right now. Keep in mind that changing these requires a master restart though, so if you set them too close to the wire, you could end up doing a bunch of restarts trying to fine-tune them. The masterside timeouts are kindof the big hammer, and usually I'd recommend setting them to 2x the expected time.

You can add a timeout.py to your repo, and have generator_script emit that as a prefix to the steps you want to run for a limited time.

See also, luci/recipes-py#3, though this wouldn't directly apply to you since you're bypassing recipes with generator_script (though if this bug was fixed, we could add another field to generator_script's output to set the value in the recipe side of things).

**Comment by anniesullie** [comment]
Thanks, Robbie! Do you have more information on the timeouts for the master? In the documentation I see builder_timeout_s: https://chromium.googlesource.com/infra/infra/+/master/doc/users/services/buildbot/builders.pyl.md#builder_timeout_s

That would allow us to set up a maximum timeout for the whole build, but if we got any new bots we'd need to make sure to add it to each individually, right?

For timeout.py, you mean this? https://code.google.com/p/chromium/codesearch#chromium/build/scripts/slave/timeout.py&q=timeout.py

So our generator would run commands like `python timeout.py mytest.py`?

**Comment by riannucci** [comment]
Yes, I think it's `builder_timeout_s`. It may be worth adding a facility to enable builders.pyl to do X seconds without output timeout too (buildbot has it, but builders.pyl doesn't seem to).

For timeout.py, yes, I think you'd want that, but let's pass the path to timeout.py to the generator script so that we don't get weird implicit dependencies. So for now, do something like:

```python
api.generator_script(path_to_script, ..., 
  "--timeout-py", api.path['build'].join('scripts', 'slave', 'timeout.py'), ...)
```

And then that will show up to your generator script as an absolute path.

**Comment by anniesullie** [comment]
Actually just got a comment on a related issue, looks like this can be solved with master-side timeouts, gonna try that: https://bugs.chromium.org/p/chromium/issues/detail?id=516784#c10

**Comment by anniesullie** [comment]
Whoops, that's the same suggestion you gave earlier. But it looks like we don't add tons of slaves, and the expected time is generally known, so I think we're going to go with it.

**Comment by riannucci** [comment]
Per-step timeouts were added in https://github.com/luci/recipes-py/commit/d6668158ac17d845f016c191258b733947169ad1 , which closed luci/recipes-py#3 .
